### ISSUE TITLE
Wrong parsing of an application name containing "." (a dot) leads to a non-working AppImage

### ISSUE THREAD

**Issue by ludo456** [probonopd/linuxdeployqt#289]
## Steps to reproduce:

1.  Create an AppDir containing _usr/bin/my_application_2.0_linux64_  and _usr/share/applications/my_application.desktop_
=> my_application_2.0_linux64 is actually the name of the executable.

2.  Run the following command line in a shell:
`./linuxdeployqt-continuous-x86_64.AppImage AppDir/usr/share/applications/my_application.desktop -appimage -bundle-non-qt-libs -verbose=2`

- [ ] You will notice that the AppDir was filled properly

- [ ] The AppImage was created without error

- [ ] **But** AppDir/AppRun is a symlink to AppDir/bin/my_application2
```
> ls -la AppDir
21 May  7 23:09 AppRun -> usr/bin/my_application2
```
 that is to say the application name was truncated at the dot (parsing error). Trying to run the AppImage yields the following message :
`execv error: No such file or directory`

## Work around
Rename the application binary file without any dot, for example _my_application_2_linux64_

**Comment by TheAssassin** [comment]
Oh well, this should not happen. I bet it's due to calling `QFile(Info)`'s `baseName` instead of `fileName` somewhere. They define both terms differently: in Qt's case, `fileName` returns the complete file name (not the full path), and `baseName` returns the file name without the file extension (which for them means, split by the first dot).

Should be an easy fix.

**Comment by probonopd** [comment]
Spot on... likely here:

https://github.com/probonopd/linuxdeployqt/blob/63e768b82580a3f10ae1461f22e4f530f038f824/tools/linuxdeployqt/main.cpp#L196

Thanks @TheAssassin - you are now a true master of Qt, debugging code without even looking at it ;-)

**Comment by TheAssassin** [comment]
Not really a master, but I have some experience with its `QFileInfo` class. Also, I have read a large portion of the Qt docs.

**Comment by tresf** [comment]
Would there be any reconsideration to add back in a check for the `.real` extension?  It was proposed to use when we worked with @probonopd to originally create our first AppImage, but this patch reverts that option for us.  More context is here: https://github.com/LMMS/lmms/issues/4515#issuecomment-410429855

**Comment by probonopd** [comment]
How would a proper fix look like?

**Comment by probonopd** [comment]
Well, hardcoding one particular extension, `.real`, is not a good idea imho. We could just as well just append some string ("real" or another) without a dot.

Can you point me to the concext

**Comment by tresf** [comment]
> Well, hardcoding one particular extension, .real, is not a good idea imho. We could just as well just append some string ("real" or another) without a dot.

I'm not sure what you mean.  If this is an officially supported technique, you should set the standard, as this is your library.  I was hesitant to even write it, TBH since this was a proposal by exactly you. :D

Whatever you decide, if you decide, let us know so that we can continuing using `.AppImage` for our project.  If we're doing it wrong, we're happy to know that too. 👍

**Comment by tresf** [comment]
We'd be happy to provide this as a supplement to the `.desktop` specification (e.g. `ExecAppImage=foo.shim`) but since we're still [volatile](#227) to a Continuous release, we can't release until we have a viable solution.

**Comment by probonopd** [comment]
Maybe I am seeing it too simple, but it seems that a dot (`.`) in the `Exec=` value breaks it. So what I am suggesting is not to use a dot. For example, have `Exec=launcher`, and have `usr/bin/launcher` as a shell script that sets the environment variables as needed and launches the main application.

Am I missing something?

**Comment by tresf** [comment]
> The easy workaround is to just not use a dot

Perhaps you are misunderstanding the bug and the resolution.  The dot in the filename **is what allowed the shim script to work in the first place**.  Adding an underscore does nothing because it was the behavior of the baseName that allowed the shim in the first place as it allowed the binary to have a different path than the launch command without breaking the automatic bundling behavior of the tool.

**Comment by probonopd** [comment]
Ah right, I forgot about how that script works. Can we change the shim script to use a `_` rather than `.`?

**Comment by tresf** [comment]
`linuxdeployqt` reads the `Exec=` line and then does it's bundling based on that.
* Previously, if someone were to move `/usr/bin/foo` to `/usr/bin/foo.real` and specify `Exec=foo.real`, the `AppRun` would automatically drop off the `.real` extension and point to the shim located at `/usr/bin/foo`.  Conveniently, the tool still would actually bundle dependencies of the executable `foo.real` but when launching the AppImage, would launch `foo`, due to the dropping of the file extension, allowing a shim script in the executable's place that handles just-in-time shimming.
* Now, if someone were to specify `Exec=foo.real`, the tool actually bundles `foo.real` which is good, but the `AppRun` also points to `foo.real`, so the shim is unused.

> Can we change the shim script to use a `_` rather than `.`?

I feel we're talking in circles here.  With your advice, we moved the binary and `linuxdeployqt` was ok with it because of this bug.  If you have no intentions of adding shim support back in, we'll have to write our own.

**Comment by TheAssassin** [comment]
You might want to check out [linuxdeploy](https://github.com/linuxdeploy/linuxdeploy) and its [Qt plugin](https://github.com/linuxdeploy/linuxdeploy-plugin-qt).

**Comment by tresf** [comment]
Since we've switched to using `appimagetool`, it looks like we can just fix this symlink ourselves before the call to `appimagetool`.

```bash
# After calling linuxdeployqt, point the AppRun to the shim launcher, not the foo.real version
ln -sfr [appdir]/usr/bin/foo [appdir]/AppRun
```

**Comment by TheAssassin** [comment]
@tresf linuxdeploy for instance respects existing AppRuns, and doesn't overwrite them.

**Comment by tresf** [comment]
Thanks, we'll eventually switch to that if it's the eventual successor to `linuxdeployqt`.  When we first integrated with `linuxdeployqt` we didn't use `appimagetool` at all so we just used the `.real` trick as proposed.  Since then we've gotten deeper into the workaround and have developed some fine-grained needs, so we're in good shape just making the `AppRun` ourselves after `linuxdeployqt` is done doing its thing.  I like the idea of rolling our own `AppRun` script and not using a symlink.  We'll eventually move to that after we get our next major release out the door.
